// trezor-connect/lib/constants/device.js
'use strict';

// device list events

exports.__esModule = true;
var CONNECT = exports.CONNECT = 'device-connect';
var CONNECT_UNACQUIRED = exports.CONNECT_UNACQUIRED = 'device-connect_unacquired';
var DISCONNECT = exports.DISCONNECT = 'device-disconnect';
var CHANGED = exports.CHANGED = 'device-changed';
var ACQUIRE = exports.ACQUIRE = 'device-acquire';
var RELEASE = exports.RELEASE = 'device-release';
var ACQUIRED = exports.ACQUIRED = 'device-acquired';
var RELEASED = exports.RELEASED = 'device-released';
var USED_ELSEWHERE = exports.USED_ELSEWHERE = 'device-used_elsewhere';

var LOADING = exports.LOADING = 'device-loading';

// trezor-link events in protobuf format
var BUTTON = exports.BUTTON = 'button';
var PIN = exports.PIN = 'pin';
var PASSPHRASE = exports.PASSPHRASE = 'passphrase';
var PASSPHRASE_ON_DEVICE = exports.PASSPHRASE_ON_DEVICE = 'passphrase_on_device';
var WORD = exports.WORD = 'word';

// custom
var WAIT_FOR_SELECTION = exports.WAIT_FOR_SELECTION = 'device-wait_for_selection';

// this string has different prefix than other constants and it's used as device path
var UNREADABLE = exports.UNREADABLE = 'unreadable-device';
