// trezor-connect/lib/constants/ui.js
'use strict';

exports.__esModule = true;
var IFRAME_HANDSHAKE = exports.IFRAME_HANDSHAKE = 'iframe-handshake';

var TRANSPORT = exports.TRANSPORT = 'ui-no_transport';
var BOOTLOADER = exports.BOOTLOADER = 'ui-device_bootloader_mode';
var INITIALIZE = exports.INITIALIZE = 'ui-device_not_initialized';
var FIRMWARE = exports.FIRMWARE = 'ui-device_firmware_old';
var FIRMWARE_OUTDATED = exports.FIRMWARE_OUTDATED = 'ui-device_firmware_outdated';
var FIRMWARE_NOT_SUPPORTED = exports.FIRMWARE_NOT_SUPPORTED = 'ui-device_firmware_unsupported';
var BROWSER_NOT_SUPPORTED = exports.BROWSER_NOT_SUPPORTED = 'ui-browser_not_supported';
var BROWSER_OUTDATED = exports.BROWSER_OUTDATED = 'ui-browser_outdated';
var RECEIVE_BROWSER = exports.RECEIVE_BROWSER = 'ui-receive_browser';

var REQUEST_UI_WINDOW = exports.REQUEST_UI_WINDOW = 'ui-request_window';
var CLOSE_UI_WINDOW = exports.CLOSE_UI_WINDOW = 'ui-close_window';

var REQUEST_PERMISSION = exports.REQUEST_PERMISSION = 'ui-request_permission';
var REQUEST_CONFIRMATION = exports.REQUEST_CONFIRMATION = 'ui-request_confirmation';
var REQUEST_PIN = exports.REQUEST_PIN = 'ui-request_pin';
var INVALID_PIN = exports.INVALID_PIN = 'ui-invalid_pin';
var REQUEST_PASSPHRASE = exports.REQUEST_PASSPHRASE = 'ui-request_passphrase';
var REQUEST_PASSPHRASE_ON_DEVICE = exports.REQUEST_PASSPHRASE_ON_DEVICE = 'ui-request_passphrase_on_device';
var INVALID_PASSPHRASE = exports.INVALID_PASSPHRASE = 'ui-invalid_passphrase';
var INVALID_PASSPHRASE_ACTION = exports.INVALID_PASSPHRASE_ACTION = 'ui-invalid_passphrase_action';
var CONNECT = exports.CONNECT = 'ui-connect';
var LOADING = exports.LOADING = 'ui-loading';
var SET_OPERATION = exports.SET_OPERATION = 'ui-set_operation';
var SELECT_DEVICE = exports.SELECT_DEVICE = 'ui-select_device';
var SELECT_ACCOUNT = exports.SELECT_ACCOUNT = 'ui-select_account';
var SELECT_FEE = exports.SELECT_FEE = 'ui-select_fee';
var UPDATE_CUSTOM_FEE = exports.UPDATE_CUSTOM_FEE = 'ui-update_custom_fee';
var INSUFFICIENT_FUNDS = exports.INSUFFICIENT_FUNDS = 'ui-insufficient_funds';
var REQUEST_BUTTON = exports.REQUEST_BUTTON = 'ui-button';

var RECEIVE_PERMISSION = exports.RECEIVE_PERMISSION = 'ui-receive_permission';
var RECEIVE_CONFIRMATION = exports.RECEIVE_CONFIRMATION = 'ui-receive_confirmation';
var RECEIVE_PIN = exports.RECEIVE_PIN = 'ui-receive_pin';
var RECEIVE_PASSPHRASE = exports.RECEIVE_PASSPHRASE = 'ui-receive_passphrase';
var RECEIVE_DEVICE = exports.RECEIVE_DEVICE = 'ui-receive_device';
var CHANGE_ACCOUNT = exports.CHANGE_ACCOUNT = 'ui-change_account';
var RECEIVE_ACCOUNT = exports.RECEIVE_ACCOUNT = 'ui-receive_account';
var RECEIVE_FEE = exports.RECEIVE_FEE = 'ui-receive_fee';

var CHANGE_SETTINGS = exports.CHANGE_SETTINGS = 'ui-change_settings';

var CUSTOM_MESSAGE_REQUEST = exports.CUSTOM_MESSAGE_REQUEST = 'ui-custom_request';
var CUSTOM_MESSAGE_RESPONSE = exports.CUSTOM_MESSAGE_RESPONSE = 'ui-custom_response';

var LOGIN_CHALLENGE_REQUEST = exports.LOGIN_CHALLENGE_REQUEST = 'ui-login_challenge_request';
var LOGIN_CHALLENGE_RESPONSE = exports.LOGIN_CHALLENGE_RESPONSE = 'ui-login_challenge_response';

var BUNDLE_PROGRESS = exports.BUNDLE_PROGRESS = 'ui-bundle_progress';
